File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.kt
class Solution {
    fun solution(numbers: IntArray, hand: String): String {
        val answer = StringBuilder()
        var leftHand = -1
        var rightHand = -1

        for (i in numbers) {
            when(i) {
                1, 4, 7 -> {
                    answer.append("L")
                    leftHand = i
                }
                3, 6, 9 -> {
                    answer.append("R")
                    rightHand = i
                }
                else -> {
                    if(fingerDistance(leftHand, i) > fingerDistance(rightHand, i)) {
                        answer.append("R")
                        rightHand = i
                    }else if (fingerDistance(leftHand, i) < fingerDistance(rightHand, i)) {
                        answer.append("L")
                        leftHand = i
                    }else {
                        if (hand == "left") {
                            answer.append("L")
                            leftHand = i
                        }else {
                            answer.append("R")
                            rightHand = i
                        }
                    }
                }
            }
        }
        return answer.toString()
    }

    fun fingerDistance(hand: Int, destination: Int): Int {
        return when(destination) {
            2 -> {
                when(hand) {
                    -1 -> 4
                    0, 7, 9 -> 3
                    4, 6, 8 -> 2
                    1, 3, 5 -> 1
                    else -> 0
                }
            }
            5 -> {
                when(hand) {
                    -1 -> 3
                    0, 1, 3, 7, 9 -> 2
                    2, 4, 6, 8 -> 1
                    else -> 0
                }
            }
            8 -> {
                when(hand) {
                    1, 3 -> 3
                    -1, 2, 4, 6 -> 2
                    0, 5, 7, 9 -> 1
                    else -> 0
                }
            }
            else -> {
                when(hand) {
                    1, 3 -> 4
                    2, 4, 6 -> 3
                    5, 7, 9 -> 2
                    -1, 8 -> 1
                    else -> 0
                }
            }
        }
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.swift
func solution(_ numbers:[Int], _ hand:String) -> String {
    var answer = ""
    var leftHand = 10
    var rightHand = 12
    
    for number in numbers {
        switch number {
        case 1, 4, 7:
            answer += "L"
            leftHand = number
        case 3, 6, 9:
            answer += "R"
            rightHand = number
        default:
            let leftDistance = (number == 0) ? getDistance(11, leftHand) : getDistance(number, leftHand)
            let rightDistance = (number == 0) ? getDistance(11, rightHand) : getDistance(number, rightHand)
            
            if leftDistance == rightDistance {
                if hand == "left" {
                    answer += "L"
                    leftHand = (number == 0) ? 11 : number
                } else {
                    answer += "R"
                    rightHand = (number == 0) ? 11 : number
                }
            } else if leftDistance > rightDistance {
                answer += "R"
                rightHand = (number == 0) ? 11 : number
            } else {
                answer += "L"
                leftHand = (number == 0) ? 11 : number
            }
        }
    }
    
    return answer
}

func getDistance(_ number: Int, _ hand: Int) -> Int {
    switch hand {
    case 1, 4, 7, 10:
        return abs((number - (hand+1)) / 3) + 1
    case 3, 6, 9, 12:
        return abs((number - (hand-1)) / 3) + 1
    default:
        return abs((number - hand) / 3)
    }
}